File: backend/app/schemas/nda.py
"""NDA-related Pydantic schemas"""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field, field_validator

from app.models.nda_signature import NDARole


class NDASignRequest(BaseModel):
    """Schema for signing an NDA"""
    full_legal_name: str = Field(
        ...,
        min_length=2,
        max_length=255,
        description="Full legal name as signature"
    )

    @field_validator('full_legal_name')
    @classmethod
    def validate_full_legal_name(cls, v: str) -> str:
        """Validate and clean the legal name"""
        v = v.strip()
        if len(v) < 2:
            raise ValueError('Full legal name must be at least 2 characters')
        # Basic validation - name should contain at least one space (first + last name)
        # Relaxed to allow single names for cultural inclusivity
        if not any(c.isalpha() for c in v):
            raise ValueError('Full legal name must contain letters')
        return v


class NDASignResponse(BaseModel):
    """Response after signing an NDA"""
    success: bool
    message: str
    signature_id: int
    signed_at: datetime
    nda_version: str
    review_request_id: int


class NDASignatureResponse(BaseModel):
    """Schema for NDA signature details"""
    id: int
    review_request_id: int
    user_id: int
    role: str
    full_legal_name: str
    nda_version: str
    signed_at: datetime

    # User info (populated from relationship)
    user_username: Optional[str] = None
    user_avatar: Optional[str] = None

    class Config:
        from_attributes = True


class NDAStatusResponse(BaseModel):
    """Schema for checking NDA status on a review request"""
    review_request_id: int
    requires_nda: bool
    nda_version: Optional[str] = None
    creator_signed: bool = False
    creator_signed_at: Optional[datetime] = None
    current_user_signed: bool = False
    current_user_signed_at: Optional[datetime] = None
    can_view_content: bool = Field(
        ...,
        description="Whether the current user can view the full request content"
    )


class NDAContentResponse(BaseModel):
    """Schema for returning NDA document content"""
    version: str
    title: str = "Non-Disclosure Agreement (NDA)"
    subtitle: str = "Between Creator and Reviewer on Critvue"
    content: str = Field(..., description="Full NDA text content")
    effective_date: Optional[str] = None

File: frontend/lib/api/nda.ts
/**
 * NDA API Client
 * Handles all NDA-related API requests
 */

import apiClient from "./client";

// NDA sign request
export interface NDASignRequest {
  full_legal_name: string;
}

// NDA sign response
export interface NDASignResponse {
  success: boolean;
  message: string;
  signature_id: number;
  signed_at: string;
  nda_version: string;
  review_request_id: number;
}

// NDA status response
export interface NDAStatusResponse {
  review_request_id: number;
  requires_nda: boolean;
  nda_version?: string;
  creator_signed: boolean;
  creator_signed_at?: string;
  current_user_signed: boolean;
  current_user_signed_at?: string;
  can_view_content: boolean;
}

// NDA content response
export interface NDAContentResponse {
  version: string;
  title: string;
  subtitle: string;
  content: string;
  effective_date?: string;
}

// NDA signature details
export interface NDASignature {
  id: number;
  review_request_id: number;
  user_id: number;
  role: "creator" | "reviewer";
  full_legal_name: string;
  nda_version: string;
  signed_at: string;
  user_username?: string;
  user_avatar?: string;
}

/**
 * Get NDA document content (public endpoint)
 */
export async function getNDAContent(): Promise<NDAContentResponse> {
  return apiClient.get<NDAContentResponse>("/nda/content");
}

/**
 * Check NDA status for a review request
 */
export async function getNDAStatus(reviewId: number): Promise<NDAStatusResponse> {
  return apiClient.get<NDAStatusResponse>(`/nda/status/${reviewId}`);
}

/**
 * Sign NDA for a review request
 */
export async function signNDA(
  reviewId: number,
  data: NDASignRequest
): Promise<NDASignResponse> {
  return apiClient.post<NDASignResponse>(`/nda/sign/${reviewId}`, data);
}

/**
 * Get all NDA signatures for a review request
 * (Owner sees all, others see only their own)
 */
export async function getNDASignatures(reviewId: number): Promise<NDASignature[]> {
  return apiClient.get<NDASignature[]>(`/nda/signatures/${reviewId}`);
}
